feat(config): validate reconfigure branch (#24699)
renovatebot/renovate: docs/usage/config-validation.md

You can create a [pre-commit](https://pre-commit.com) hook to validate your configuration automatically.
Go to the [`renovatebot/pre-commit-hooks` repository](https://github.com/renovatebot/pre-commit-hooks) for more information.

### Validation of Renovate config change PRs

> Renovate can validate configuration changes in Pull Requests as long as you use a designated branch name.

Reviewer: ```suggestion
Renovate can validate configuration changes in Pull Requests when you use a special branch name.
```

Follow these steps to validate your configuration:

1. Create a new Git branch that matches the `{{branchPrefix}}reconfigure` pattern. For example, if you're using the default prefix `renovate/`, your branch name must be `renovate/reconfigure`.
1. Commit your updated Renovate config file to this branch, and push it to your Git hosting platform.

> Renovate will now:

Reviewer: Good catch! 😄

> Follow these steps:
> 
> 1. Create a new Git branch that matches the `{{branchPrefix}}reconfigure` pattern.
>    For example, if you're using the default prefix `renovate/`, your branch name must be `renovate/reconfigure`. Commit your updated Renovate config file to this branch, and push it to your Git hosting platform.
> 
> 2. Each time Renovate runs, it searches for a branch that matches the reconfigure pattern and validates the Renovate configuration file.
> 
> 3. Renovate will check for a config file in the reconfigure branch, even if it has been renamed compared to the existing config file in the default branch.
> 
> 4. Depending on the outcome of the Renovate config validation run, Renovate will add a passing or failing check to the branch.
> 
> 5. If there's an open pull request from the _reconfigure_ branch then Renovate will comment in that PR with details in case there are validation errors. Each commit will be validated again the next time Renovate runs on the repository.

Reviewer: ```suggestion
Follow these steps to validate your configuration:

1. Create a new Git branch that matches the `{{branchPrefix}}reconfigure` pattern. For example, if you're using the default prefix `renovate/`, your branch name must be `renovate/reconfigure`.
1. Commit your updated Renovate config file to this branch, and push it to your Git hosting platform.

Renovate will now:

1. Search for a branch that matches the special reconfigure pattern.
1. Check for a config file in the reconfigure branch. Renovate can even find a renamed configuration file (compared to the config file in the default branch).
1. Add a passing or failing status to the branch, depending on the outcome of the config validation run.
1. If there's an _open_ pull request with validation errors from the _reconfigure_ branch then Renovate comments in the PR with details.
1. Validate each commit the next time Renovate runs on the repository, until the PR is merged.
```

We say: `Follow these steps` but only step 1 is for the user, the rest explains what Renovate will do afterwards. So let's split things up into a section for the user, and then step through what Renovate does.

You can create a [pre-commit](https://pre-commit.com) hook to validate your configuration automatically.
Go to the [`renovatebot/pre-commit-hooks` repository](https://github.com/renovatebot/pre-commit-hooks) for more information.

### Let Renovate check your config file as you're writing it

After using Renovate for a while you likely want to change Renovate's config.
Renovate can check your new config as you're writing it, but you must put your work in a branch with a _special_ name.

Follow these steps:

1. Create a new Git branch that matches the `{{branchPrefix}}reconfigure` pattern.
   For example, if you're using the default prefix `renovate/`, your branch name must be `renovate/reconfigure`. Commit your updated Renovate config file to this branch, and push it to your Git hosting platform.

2. Each time Renovate runs, it searches for a branch that matches the reconfigure pattern and validates the Renovate configuration file.

> 3. Renovate will check the config file on the reconfigure branch, even if it has a different name, compared to the config file on the default branch.

Reviewer 2: ```suggestion
3. Renovate will check for a config file in the reconfigure branch, even if it has been renamed compared to the existing config file in the default branch.
```

>    So feel free to change the Renovate config file name.

Reviewer 2: ```suggestion
```

> ### Reconfigure Renovate Config
> 
> To enable validation of your Renovate configuration changes, follow these steps:
> 
> 1. Ensure that the branch name includes the prefix `{{branchPrefix}}reconfigure`.
>    For example, if you're using the default prefix `renovate/`, your branch name should be `renovate/reconfigure`.
> 
> 2. During each Renovate run, the bot searches for a branch with the reconfigure pattern and validates the associated configuration file.
> 
> 3. Renovate does not assume that the config file has the same name as the default branch.
>    You can also modify the config file name if needed.
> 
> 4. If the configuration validation passes, Renovate adds a successful status check to the branch.
>    In the event of validation failure, a failing status check is added.
> 
> 5. Additionally, if there is an open pull request with the same branch, Renovate will post a comment on the PR listing all the validation errors for your reference.
> 
> These steps ensure that your Renovate configuration changes are validated effectively, helping you maintain a reliable and error-free setup.

Reviewer: ```suggestion
### Let Renovate check your config file as you're writing it

After using Renovate for a while you likely want to change Renovate's config.
Renovate can check your new config as you're writing it, but you must put your work in a branch with a _special_ name.

Follow these steps:

1. Create a new Git branch that matches the `{{branchPrefix}}reconfigure` pattern.
   For example, if you're using the default prefix `renovate/`, your branch name must be `renovate/reconfigure`. Commit your updated Renovate config file to this branch, and push it to your Git hosting platform.

2. Each time Renovate runs, it searches for a branch that matches the reconfigure pattern and validates the Renovate configuration file.

3. Renovate will check the config file on the reconfigure branch, even if it has a different name, compared to the config file on the default branch.
   So feel free to change the Renovate config file name.

4. Depending on the outcome of the Renovate config validation run, Renovate will add a passing or failing check to the branch.

5. If there's an open pull request from the _reconfigure_ branch then Renovate will comment in that PR and list all the validation errors. Commit and push your fixes to the branch, and try again. When the config validation passes, you can merge your Renovate config.
```

How about something like this? This is a very cool new feature, so we should write about it in clear words, and explain how the user benefits from it. Check if my rewrite is still technically correct (the best kind of correct 🙃).

Also does the "check my config on the special branch"-feature use the `--strict` mode? It's not listed in the current docs... 😉 

When we're finished hashing out my rewrite, we should think about moving the whole section nearer to the top. I guess most users will want to use the "create special branch to let Renovate check my config"-feature instead of dealing with installing the npm package, and running the validator locally.

So the final flow of the document is more like:

- What's config validation
- Different ways to validate config (manually, special branch feature)
- Special branch feature explanation
- Manual validation

Author: Thanks! for the refactor, it sounds better. ❤️

The strict mode doesn't apply to this feature.

The order is correct as well.

> 5. If there's an open pull request from the _reconfigure_ branch then Renovate will comment in that PR and list all the validation errors. Commit and push your fixes to the branch, and try again. When the config validation passes, you can merge your Renovate config.

Reviewer 2: ```suggestion
5. If there's an open pull request from the _reconfigure_ branch then Renovate will comment in that PR with details in case there are validation errors. Each commit will be revalidated the next time Renovate runs on the repository.
```

2. Each time Renovate runs, it searches for a branch that matches the reconfigure pattern and validates the Renovate configuration file.

3. Renovate will check for a config file in the reconfigure branch, even if it has been renamed compared to the existing config file in the default branch.

4. Depending on the outcome of the Renovate config validation run, Renovate will add a passing or failing check to the branch.

> 5. If there's an open pull request from the _reconfigure_ branch then Renovate will comment in that PR with details in case there are validation errors. Each commit will be revalidated the next time Renovate runs on the repository.

Reviewer: ```suggestion
5. If there's an open pull request from the _reconfigure_ branch then Renovate will comment in that PR with details in case there are validation errors. Each commit will be validated again the next time Renovate runs on the repository.
```
